pennylane/devices/capabilities.py

```diff
@@ -0,0 +1,313 @@
+
+    FINITE_SHOTS_ONLY = "finiteshots"
+    """If the operator or measurement process is only supported with finite shots."""
+
+    ANALYTIC_MODE_ONLY = "analytic"
+    """If the operator or measurement process is only supported in analytic execution."""
+
+    TERMS_MUST_COMMUTE = "terms-commute"
+    """If the composite operator is supported only when its terms commute."""
+
+
+VALID_CONDITION_STRINGS = {condition.value for condition in ExecutionCondition}
+
+
+@dataclass
+class OperatorProperties:
+    """Information about support for each operation.
+
+    Attributes:
+        invertible (bool): Whether the adjoint of the operation is also supported.
+        controllable (bool): Whether the operation can be controlled.
+        differentiable (bool): Whether the operation is supported for device gradients.
+        conditions (list[ExecutionCondition]): Execution conditions that the operation must meet.
+    """
+
+    invertible: bool = False
+    controllable: bool = False
+    differentiable: bool = False
+    conditions: list[ExecutionCondition] = field(default_factory=list)
+
+
+@dataclass
+class ProgramFeatures:
+    """Program features, obtained from the user.
+
+    This is used to filter the device capabilities based on their conditions. Currently, the
+    only supported filter is whether the execution mode is finite shots.
+    """
+
+    finite_shots: bool
+
+
+@dataclass
+class DeviceCapabilities:  # pylint: disable=too-many-instance-attributes
+    """Capabilities of a quantum device.
+
+    Attributes:
+        operations: Operations natively supported by the backend device.
+        observables: Observables that the device can measure.
+        measurement_processes: List of measurement processes supported by the backend device.
+        qjit_compatible (bool): Whether the device is compatible with qjit.
+        mid_circuit_measurements (bool): Whether the device supports mid-circuit measurements natively.
+        runtime_code_generation (bool): Whether the device requires run time generation of the quantum circuit.
+        dynamic_qubit_management (bool): Whether the device supports dynamic qubit allocation/deallocation.
+        overlapping_observables (bool): Whether the device supports measuring overlapping observables on the same tape.
+        non_commuting_observables (bool): Whether the device supports measuring non-commuting observables on the same tape.
+        initial_state_prep (bool): Whether the device supports initial state preparation.
+    """
+
+    operations: dict[str, OperatorProperties] = field(default_factory=dict)
+    observables: dict[str, OperatorProperties] = field(default_factory=dict)
+    measurement_processes: dict[str, list[ExecutionCondition]] = field(default_factory=dict)
+    qjit_compatible: bool = False
+    mid_circuit_measurements: bool = False
+    runtime_code_generation: bool = False
+    dynamic_qubit_management: bool = False
+    overlapping_observables: bool = True
```

Comment:
`overlapping_observables` is by default `True` for consistency in behaviour. The current logic uses `qwc` grouping by default, which assumes that a device supports observables on overlapping wires as long as they commute. To enable `wires` based grouping, the plugin developer needs to explicitly set `overlapping_observables` to `False`.